resolwe/flow/executors/init_container.py

```diff
@@ -0,0 +1,210 @@
+            await download_to_location(
+                base_url, files, connectors[connector_name], inputs_connector
+            )
+    except Exception as e:
+        error_message = f"Preparing input for url {base_url} from connector {connector_name} failed: {e}."
+        await communicator.send_command(
+            Message.command("process_log", {"error": error_message})
+        )
+        await communicator.send_command(Message.command("finish", {}))
+        raise DataTransferError(error_message)
+
+
+async def download_to_location(
+    base_url: str,
+    files: dict,
+    from_connector: BaseStorageConnector,
+    to_connector: BaseStorageConnector,
+):
+    """Download missing paths.
+
+    :raises DataTransferError: on failure.
+    """
+    try:
+        logger.debug(f"Transfering from: {from_connector} --> {to_connector}.")
+        t = Transfer(from_connector, to_connector)
+        t.transfer_objects(base_url, files)
+    except DataTransferError:
+        logger.exception(
+            "Data transfer error downloading data to base url {}.".format(base_url)
+        )
+
+
+def set_permissions():
+    """Set permissions."""
+    logger.debug("Setting permissions.")
+    local_directory = DATA_LOCAL_VOLUME / LOCATION_SUBPATH
+    local_directory.mkdir()
+    shutil.chown(local_directory, GENIALIS_UID, GENIALIS_GID)
+
+
+def _get_communicator() -> ZMQCommunicator:
+    """Connect to the listener."""
+    zmq_context = zmq.asyncio.Context.instance()
+    zmq_socket = zmq_context.socket(zmq.DEALER)
+    zmq_socket.setsockopt(zmq.IDENTITY, str(DATA_ID).encode())
+    connect_string = f"{LISTENER_PROTOCOL}://{LISTENER_IP}:{LISTENER_PORT}"
+    logger.debug("Opening connection to %s", connect_string)
+    zmq_socket.connect(connect_string)
+    return ZMQCommunicator(zmq_socket, "init_container <-> listener", logger)
+
+
+class InitProtocol(BaseProtocol):
+    """Protocol class."""
+
+    def __init__(
+        self,
+        communicator: BaseCommunicator,
+        logger: logging.Logger,
+    ):
+        """Initialization."""
+        super().__init__(communicator, logger)
+        self._transfering_task: Optional[asyncio.Task] = None
+        self._terminated = False
+
+    async def post_terminate(self, message: Message, identity: PeerIdentity):
+        """Handle post-terminate command."""
+        logger.debug("Post terminate.")
+        self._terminated = True
+        if self._transfering_task is not None:
+            self._transfering_task.cancel()
+
+    async def transfer_data(self):
+        """Transfer missing data.
+
+        Transfer missing data depends on whether data is downloaded to a shared
+        filesystem or not.
+
+        :raises Exception: on error.
+        """
+        # Notify listener the data is being transfered.
+        await self.communicator.send_command(Message.command("update_status", "PP"))
+        if DATA_ALL_VOLUME_SHARED:
```

Comment:
It would, but multiple workers downloading the data for the same data object to the same location could make a mess. Also if we want to reuse the data the listener must create a StorageLocation etc. so the procedure is not entirely the same.